File: python/src/skills_utils/agents.py
from __future__ import annotations

import subprocess
from pathlib import Path
from typing import Any


def expand_path(value: str, cwd: Path | None = None) -> Path:
    base = cwd or Path.cwd()
    path = Path(value).expanduser()
    if not path.is_absolute():
        path = (base / path).resolve()
    return path


def project_root(cwd: Path | None = None) -> Path:
    base = cwd or Path.cwd()
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=base,
            capture_output=True,
            text=True,
            check=True,
        )
        return Path(result.stdout.strip())
    except (FileNotFoundError, subprocess.CalledProcessError):
        return base


def resolve_skills_dir(agent: dict[str, Any], cwd: Path | None = None) -> Path:
    base = cwd or Path.cwd()
    if agent.get("scope") == "project":
        root = project_root(base)
        return expand_path(agent["skills_dir"], root)
    return expand_path(agent["skills_dir"], base)


def list_agents(config: dict[str, Any]) -> list[tuple[str, dict[str, Any]]]:
    return list((agent_id, data) for agent_id, data in config.get("agents", {}).items())


def agent_label(agent_id: str, agent: dict[str, Any]) -> str:
    label = agent.get("label", agent_id)
    return f"{label} ({agent_id})"

File: node/src/agents.ts
import os from "node:os";
import path from "node:path";

export function expandPath(value: string, cwd = process.cwd()): string {
  const expanded = value.startsWith("~") ? path.join(os.homedir(), value.slice(1)) : value;
  return path.isAbsolute(expanded) ? expanded : path.resolve(cwd, expanded);
}

export async function projectRoot(cwd = process.cwd()): Promise<string> {
  // Project-scoped installs should target the directory where the user runs the CLI.
  return cwd;
}

export async function resolveSkillsDir(agent: any, cwd = process.cwd()): Promise<string> {
  if (agent.scope === "project") {
    const root = await projectRoot(cwd);
    return expandPath(agent.skills_dir, root);
  }
  return expandPath(agent.skills_dir, cwd);
}

export function listAgents(config: any): Array<[string, any]> {
  return Object.entries(config.agents ?? {});
}
